internal/internal_decision_state_machine.go

```diff
@@ -472,55 +473,24 @@ func (d *childWorkflowDecisionStateMachine) getDecision() *s.Decision {
 		decision := createNewDecision(s.DecisionTypeStartChildWorkflowExecution)
 		decision.StartChildWorkflowExecutionDecisionAttributes = d.attributes
 		return decision
-	case decisionStateCanceledAfterStarted:
-		decision := createNewDecision(s.DecisionTypeRequestCancelExternalWorkflowExecution)
-		decision.RequestCancelExternalWorkflowExecutionDecisionAttributes = &s.RequestCancelExternalWorkflowExecutionDecisionAttributes{
-			Domain:     d.attributes.Domain,
-			WorkflowId: d.attributes.WorkflowId,
-			RunId:      common.StringPtr(d.runID),
-		}
-		return decision
 	default:
 		return nil
 	}
 }
 
-func (d *childWorkflowDecisionStateMachine) handleDecisionSent() {
-	switch d.state {
-	case decisionStateCanceledAfterStarted:
-		d.moveState(decisionStateCancellationDecisionSent, eventDecisionSent)
-	default:
-		d.decisionStateMachineBase.handleDecisionSent()
-	}
-}
-
 func (d *childWorkflowDecisionStateMachine) handleStartedEvent() {
 	switch d.state {
 	case decisionStateInitiated:
 		d.moveState(decisionStateStarted, eventStarted)
-	case decisionStateCanceledAfterInitiated:
-		d.moveState(decisionStateCanceledAfterStarted, eventStarted)
 	default:
 		d.decisionStateMachineBase.handleStartedEvent()
 	}
 }
 
-func (d *childWorkflowDecisionStateMachine) handleCancelFailedEvent() {
-	switch d.state {
-	case decisionStateCancellationDecisionSent:
-		d.moveState(decisionStateStarted, eventCancelFailed)
-	default:
-		d.decisionStateMachineBase.handleCancelFailedEvent()
-	}
-}
-
 func (d *childWorkflowDecisionStateMachine) cancel() {
-	switch d.state {
-	case decisionStateStarted:
-		d.moveState(decisionStateCanceledAfterStarted, eventCancel)
-	default:
-		d.decisionStateMachineBase.cancel()
-	}
+	// the child workflow state machine should be alterd when trying to
+	// cancel child workflow, should use requestCancelExternalWorkflow instead
+	panic("unsupported operation")
```

Comment:
what happen for below code:
childCtx, childCancel := workflow.WithCancel(ctx)
workflow.ExecuteChildWorkflow(childCtx, ChildWorkflow, ...)
childCancel()

in old behavior, it will cancel the state machine and not even send out decision to start child.
in this new behavior, it will schedule the child even if cancel is already requested, and then wait until it started then send decision to cancel it.

This is breaking change.

Could you double check with Max.
cc @mfateev

Reply:
sure, i think i and maxim had a conversation about this, the conclusion was to record every events, so no action will be omitted

Reply:
@yiminc  talked to @mfateev, the change of behavior is not a huge issue, the issue is actually, when doing replay, the new client will encounter missing events, causing the child workflow future Get() stuck.